### PR TITLE
jbide-26159: Server adapter: port forwarding isn't stopped if you delete a pod

### PR DESCRIPTION
**Pull request by bdshadow**
Signed-off-by: Dmitrii Bocharov <dbocharo@redhat.com>

### CODE REVIEW

**Comment by bdshadow** [comment]
@jeffmaury 

-  it's not about debugging, it's more about port forwarding - look at the issue.
-  yes, that's what Andre said about, working on it. It'll be triggered on a message from the server (through the Watcher)

**Comment by jeffmaury** [comment]
OK, I got confused by the title of this PR 👎

**Comment by bdshadow** [comment]
@jeffmaury it has been changed recently by @adietish according to the description of the jira, as the title really didn't reflect the main idea

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java [hunk 1]
```diff
@@ -151,9 +157,23 @@ protected void runOCBinary() {
...
 		}, null);
-		if (portForwarding != null) {
-			REGISTRY.put(pod, portForwarding);
-		}
+        if (portForwarding != null) {
+            REGISTRY.put(pod, portForwarding);
+            IConnectionsRegistryListener podListener = new ConnectionsRegistryAdapter() {
+                @Override
+                public void connectionChanged(IConnection connection, String property, Object oldValue, Object newValue) {
+                    if (newValue == null && oldValue instanceof IPod) {
```
Comment (jeffmaury) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java]: Where do you check that the deleted pod is the one you started a port forward ?

Comment (adietish) [reply plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java]: I guess something like ```REGISTRY.containsKey(pod)``` is missing.
@bdshadow WDYT?

Comment (adietish) [reply plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java]: the following does the trick for me:
```
+++ b/plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java
@@ -162,7 +162,9 @@ public class PortForwardingUtils {
             IConnectionsRegistryListener podListener = new ConnectionsRegistryAdapter() {
                 @Override
                 public void connectionChanged(IConnection connection, String property, Object oldValue, Object newValue) {
-                    if (newValue == null && oldValue instanceof IPod) {
+                    if (newValue == null
+                               && oldValue instanceof IPod
+                               && REGISTRY.containsKey(oldValue)) {
                         try {
```

steps:
1) ASSERT: have 2 applications in 2 different projects with at least 1 pod runnin
2) EXEC: start port forwarding for both pods (2nd one needs to be started by checking "[x] find free local ports"
3) EXEC: in cmd line, verify that 2 "oc port-forwarding" processes (see exact semantic above) are running
4) EXEC: kill 1 pod
5) ASSERT: the port forwarding for the pod, that you just killed, was terminated, the other port port forwarding is still running

Comment (bdshadow) [reply plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java]: @jeffmaury @adietish if you look inside `PortForwardingUtils.stopPortForwarding((IPod)oldValue, null);`, you'll find that first of all there is a check `PortForwardingUtils.isPortForwardingStarted(pod)`, which actually checks if a pod was put into `REGISTRY` before

Comment (bdshadow) [reply plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java]: @adietish checked your steps. Everything works for me without additional check ` && REGISTRY.containsKey(oldValue)`. As i said above, this check already exists inside `PortForwardingUtils.isPortForwardingStarted(pod)`

Comment (adietish) [reply plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java]: @bdshadow yes I think that you're right. Resolving. @jeffmaury dare to comment please?

Comment (jeffmaury) [reply plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/portforwarding/PortForwardingUtils.java]: +1

**Pull request merged**